### PR TITLE
Add test for PostgresServerResource.WithInitBindMount

### PR DESCRIPTION
**Pull request by sebastienros**
Realized it was missing when I extracted MySql which has the same one.

 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/4926)

### CODE REVIEW

File: tests/Aspire.Hosting.PostgreSQL.Tests/PostgresFunctionalTests.cs [hunk 1]
```diff
@@ -229,6 +229,80 @@ public async Task WithDataShouldPersistStateBetweenUsages(bool useVolume)
...
+        try
+        {
+            if (Directory.Exists(bindMountPath))
+            {
+                Directory.Delete(bindMountPath);
+            }
+
+            Directory.CreateDirectory(bindMountPath);
+
+            File.WriteAllText(Path.Combine(bindMountPath, "init.sql"), "CREATE TABLE cars (brand VARCHAR(255)); INSERT INTO cars (brand) VALUES ('BatMobile'); SELECT * FROM cars;");
```
Comment (eerhardt) [inline tests/Aspire.Hosting.PostgreSQL.Tests/PostgresFunctionalTests.cs:254]:
```suggestion
            File.WriteAllText(Path.Combine(bindMountPath, "init.sql"), "CREATE TABLE cars (brand VARCHAR(255)); INSERT INTO cars (brand) VALUES ('BatMobile');");
```

Why do we need to select them out?

File: tests/Aspire.Hosting.PostgreSQL.Tests/PostgresFunctionalTests.cs [hunk 2]
```diff
@@ -229,6 +229,80 @@ public async Task WithDataShouldPersistStateBetweenUsages(bool useVolume)
...
+            await pipeline.ExecuteAsync(async token =>
+            {
+                using var connection = host.Services.GetRequiredService<NpgsqlConnection>();
+                await connection.OpenAsync(token);
+
+                var command = connection.CreateCommand();
+                command.CommandText = $"SELECT * FROM cars;";
+                var results = await command.ExecuteReaderAsync(token);
+
+                Assert.True(results.HasRows);
```
Comment (eerhardt) [inline tests/Aspire.Hosting.PostgreSQL.Tests/PostgresFunctionalTests.cs:290]: Can we make this stronger and assert there is 1 row?

File: tests/Aspire.Hosting.PostgreSQL.Tests/PostgresFunctionalTests.cs [hunk 3]
```diff
@@ -229,6 +229,80 @@ public async Task WithDataShouldPersistStateBetweenUsages(bool useVolume)
...
+
+        try
+        {
+            if (Directory.Exists(bindMountPath))
+            {
+                Directory.Delete(bindMountPath);
+            }
+
+            Directory.CreateDirectory(bindMountPath);
+
```
Comment (eerhardt) [inline tests/Aspire.Hosting.PostgreSQL.Tests/PostgresFunctionalTests.cs:253]:
```suggestion
        var bindMountPath = Directory.CreateTempSubdirectory();

        try
        {
```

This API is guaranteed to create a new, empty directory.

**Review by eerhardt — APPROVED** [review]
Looks good. Just some minor tweaks.

**Review by mitchdenny — APPROVED** [review]

File: tests/Aspire.Hosting.PostgreSQL.Tests/PostgresFunctionalTests.cs [hunk 4]
```diff
@@ -229,6 +226,76 @@ public async Task WithDataShouldPersistStateBetweenUsages(bool useVolume)
...
+            await pipeline.ExecuteAsync(async token =>
+            {
+                using var connection = host.Services.GetRequiredService<NpgsqlConnection>();
+                await connection.OpenAsync(token);
+
+                var command = connection.CreateCommand();
+                command.CommandText = $"SELECT * FROM cars;";
+                var results = await command.ExecuteReaderAsync(token);
+
+                Assert.Single(results);
```
Comment (radical) [inline tests/Aspire.Hosting.PostgreSQL.Tests/PostgresFunctionalTests.cs:283]: Would it be useful to assert the contents of the result too?

**Comment by radical** [comment]
Test failing with:
```
Npgsql.NpgsqlException : Exception while reading from stream
---- System.TimeoutException : Timeout during reading attempt
CallStack
   at Npgsql.Internal.NpgsqlReadBuffer.<Ensure>g__EnsureLong|55_0(NpgsqlReadBuffer buffer, Int32 count, Boolean async, Boolean readingNotifications)
   at System.Runtime.CompilerServices.PoolingAsyncValueTaskMethodBuilder`1.StateMachineBox`1.System.Threading.Tasks.Sources.IValueTaskSource.GetResult(Int16 token)
   at Npgsql.Internal.NpgsqlConnector.RawOpen(SslMode sslMode, NpgsqlTimeout timeout, Boolean async, CancellationToken cancellationToken, Boolean isFirstAttempt)
   at Npgsql.Internal.NpgsqlConnector.<Open>g__OpenCore|213_1(NpgsqlConnector conn, SslMode sslMode, NpgsqlTimeout timeout, Boolean async, CancellationToken cancellationToken, Boolean isFirstAttempt)
   at Npgsql.Internal.NpgsqlConnector.Open(NpgsqlTimeout timeout, Boolean async, CancellationToken cancellationToken)
   at Npgsql.PoolingDataSource.OpenNewConnector(NpgsqlConnection conn, NpgsqlTimeout timeout, Boolean async, CancellationToken cancellationToken)
   at Npgsql.PoolingDataSource.<Get>g__RentAsync|34_0(NpgsqlConnection conn, NpgsqlTimeout timeout, Boolean async, CancellationToken cancellationToken)
   at Npgsql.NpgsqlConnection.<Open>g__OpenAsync|42_0(Boolean async, CancellationToken cancellationToken)
   at Aspire.Hosting.PostgreSQL.Tests.PostgresFunctionalTests.<>c__DisplayClass4_0.<<VerifyWithInitBindMount>b__0>d.MoveNext() in /_/tests/Aspire.Hosting.PostgreSQL.Tests/PostgresFunctionalTests.cs:line 277
--- End of stack trace from previous location ---
   at Polly.ResiliencePipeline.<>c.<<ExecuteAsync>b__3_0>d.MoveNext()
--- End of stack trace from previous location ---
   at Polly.Outcome`1.GetResultOrRethrow()
   at Polly.ResiliencePipeline.ExecuteAsync(Func`2 callback, CancellationToken cancellationToken)
   at Aspire.Hosting.PostgreSQL.Tests.PostgresFunctionalTests.VerifyWithInitBindMount() in /_/tests/Aspire.Hosting.PostgreSQL.Tests/PostgresFunctionalTests.cs:line 274
--- End of stack trace from previous location ---
----- Inner Stack Trace -----
```

And the log has this:
```
Aspire.Hosting.DistributedApplication Information: Aspire version: 8.1.0-ci
Aspire.Hosting.DistributedApplication Information: Distributed application starting.
Aspire.Hosting.DistributedApplication Information: Application host directory is: /mnt/vss/_work/1/s/tests/Aspire.Hosting.Tests
Aspire.Hosting.Dcp.DcpHostService Information: Starting DCP with arguments: start-apiserver --monitor 31344 --detach --kubeconfig "/tmp/aspire.3Rh1Ng/kubeconfig"
Aspire.Hosting.Dcp.start-apiserver.api-server Information: Starting API server...
Aspire.Hosting.Dcp.start-apiserver.api-server Information: API server started	{"Address": "127.0.0.1", "Port": 40475}
Aspire.Hosting.Dcp.start-apiserver.dcp-host Information: Starting DCP controller host
Aspire.Hosting.Dcp.start-apiserver.dcp-host Information: Started all services	{"count": 1}
Aspire.Hosting.Dcp.dcpctrl.IdeExecutableRunner Information: Executables cannot be started via IDE: missing required environment variable 'DEBUG_SESSION_PORT'
Aspire.Hosting.Dcp.dcpctrl Information: starting controller manager
Aspire.Hosting.DistributedApplication Information: Distributed application started. Press Ctrl+C to shut down.
Aspire.Hosting.Dcp.dcpctrl.ContainerReconciler
	Error: could not determine host address and port for container port	{"Container": {"name":"pg-cudsemnp-ac4b991"}, "Reconciliation": 5, "error": "container '/pg-cudsemnp-ac4b991' is not running: exited"}
Aspire.Hosting.Dcp.dcpctrl.ContainerReconciler
	Error: could not create Endpoint object	{"Container": {"name":"pg-cudsemnp-ac4b991"}, "Reconciliation": 5, "ServiceName": "pg-ac4b991", "Workload": "/pg-cudsemnp-ac4b991", "error": "container '/pg-cudsemnp-ac4b991' is not running: exited"}
```

**Comment by sebastienros** [comment]
@radical do you know if there is a way to get the Docker logs for the failure? I want to see if there is an issue during the initialization of the image, since we are passing a custom bind mount and the DCP errors seems to indicate that the container hasn't started for this test only.

**Comment by mitchdenny** [comment]
https://github.com/dotnet/aspire/blob/main/tests/Aspire.Hosting.Testing.Tests/ResourceLoggerForwarderServiceTests.cs

**Comment by sebastienros** [comment]
Worked like a charm, so I know what's happening now:

```
84: 2024-07-18T09:38:12.907160440Z ls: cannot open directory '/docker-entrypoint-initdb.d/': Permission denied

```

**Pull request merged**